cpp/oneapi/dal/algo/basic_statistics/backend/gpu/compute_kernel_csr_impl_dpc.cpp

```diff
@@ -0,0 +1,275 @@
+    auto allreduce_event =
+        results.assign_from_host(q, host_results.get_data(), res_opt_count_ * column_count);
+
+    auto final_event = q.submit([&](sycl::handler& cgh) {
+        cgh.depends_on({ allreduce_event });
+        cgh.parallel_for(distr_range, [=](std::int64_t col_idx) {
+            auto mean_val = result_ptr[stat::mean * column_count + col_idx];
+            result_ptr[stat::variance * column_count + col_idx] =
+                result_ptr[stat::sum2_cent * column_count + col_idx] / (row_count - 1);
+            result_ptr[stat::stddev * column_count + col_idx] =
+                sycl::sqrt(result_ptr[stat::variance * column_count + col_idx]);
+            result_ptr[stat::variation * column_count + col_idx] =
+                result_ptr[stat::stddev * column_count + col_idx] / mean_val;
+        });
+    });
+
+    final_event.wait_and_throw();
+}
+
+template <typename Float>
+result_t compute_kernel_csr_impl<Float>::operator()(const bk::context_gpu& ctx,
+                                                    const descriptor_t& desc,
+                                                    const input_t& input) {
+    auto queue = ctx.get_queue();
+    const auto table = input.get_data();
+    ONEDAL_ASSERT(table.get_kind() == csr_table::kind());
+    const csr_table csr_tdata = static_cast<const csr_table&>(table);
+    comm_t comm = ctx.get_communicator();
+    const bool distr_mode = comm.get_rank_count() > 1;
+    const auto column_count = csr_tdata.get_column_count();
+    const auto row_count = csr_tdata.get_row_count();
+    auto result_options = desc.get_result_options();
+    const auto nonzero_count = csr_tdata.get_non_zero_count();
+    auto [csr_data, column_indices, row_offsets] =
+        csr_accessor<const Float>(csr_tdata).pull(queue,
+                                                  { 0, -1 },
+                                                  sparse_indexing::zero_based,
+                                                  sycl::usm::alloc::device);
+    auto csr_data_ptr = csr_data.get_data();
+    auto column_indices_ptr = column_indices.get_data();
+
+    using limits_t = std::numeric_limits<Float>;
+    constexpr Float maximum = limits_t::max();
+
+    auto result_data = pr::ndarray<Float, 2>::empty(queue,
+                                                    { res_opt_count_, column_count },
+                                                    sycl::usm::alloc::device);
+    auto result_data_ptr = result_data.get_mutable_data();
+
+    const auto local_size =
+        bk::device_max_wg_size(queue); // number of columns processed in one group
+
+    const auto nd_range = bk::make_multiple_nd_range_1d(local_size, local_size);
+    auto event = queue.submit([&](sycl::handler& cgh) {
+        sycl::local_accessor<Float, 1> local_res_buf(local_size * res_opt_count_, cgh);
+        sycl::local_accessor<std::int32_t, 1> local_hist_buf(local_size, cgh);
+        cgh.parallel_for(nd_range, [=](auto item) {
```

Comment:
Please add a comment(s) with high-level description what happens in this parallel_for
I.e. each work item processes it's set of columns in the input matrix, the local basic statistics are updated atomically, ...
Not all the basic statistics are updated in the main loop because...